### ISSUE TITLE
Retry count is not decremented for failing job

### ISSUE THREAD

**Issue by fnothaft** [BD2KGenomics/toil#695]
I experienced a weird case today where a job would hang on startup. When I killed the job (since it was hung; I killed it manually via `ps aux | grep toil` to get the pid and `kill <pid>`), the failure count would be reported as decremented... to 1, repeatedly:

```
WARNING:toil.batchSystems.mesos.batchSystem:Task 31 failed with exit status -15
WARNING:toil.leader:Batch system is reporting that the jobWrapper with batch system ID: 31 and jobWrapper store ID: 3b2e3ae6-7a46-4686-99fa-5b0b7dc02eec failed with exit value -15
WARNING:toil.leader:No log file is present, despite jobWrapper failing: 3b2e3ae6-7a46-4686-99fa-5b0b7dc02eec
WARNING:toil.jobWrapper:Due to failure we are reducing the remaining retry count of job 3b2e3ae6-7a46-4686-99fa-5b0b7dc02eec to 1
INFO:toil.batchSystems.mesos.batchSystem:Preparing to launch Mesos task 45 using offer d05e5caa-2e5c-44cb-a1d1-79ce6cec9eb5-O228155...
INFO:toil.batchSystems.mesos.batchSystem:...launching Mesos task 45
INFO:toil.leader:Reissued any over long jobs
INFO:toil.leader:Rescued any (long) missing jobs
INFO:toil.leader:Reissued any over long jobs
INFO:toil.leader:Rescued any (long) missing jobs
WARNING:toil.batchSystems.mesos.batchSystem:Task 45 failed with exit status 143
WARNING:toil.leader:Batch system is reporting that the jobWrapper with batch system ID: 45 and jobWrapper store ID: 3b2e3ae6-7a46-4686-99fa-5b0b7dc02eec failed with exit value 143
WARNING:toil.leader:No log file is present, despite jobWrapper failing: 3b2e3ae6-7a46-4686-99fa-5b0b7dc02eec
WARNING:toil.jobWrapper:Due to failure we are reducing the remaining retry count of job 3b2e3ae6-7a46-4686-99fa-5b0b7dc02eec to 1
INFO:toil.batchSystems.mesos.batchSystem:Preparing to launch Mesos task 46 using offer d05e5caa-2e5c-44cb-a1d1-79ce6cec9eb5-O231848...
INFO:toil.batchSystems.mesos.batchSystem:...launching Mesos task 46
WARNING:toil.batchSystems.mesos.batchSystem:Task 46 failed with exit status -15
WARNING:toil.leader:Batch system is reporting that the jobWrapper with batch system ID: 46 and jobWrapper store ID: 3b2e3ae6-7a46-4686-99fa-5b0b7dc02eec failed with exit value -15
WARNING:toil.leader:No log file is present, despite jobWrapper failing: 3b2e3ae6-7a46-4686-99fa-5b0b7dc02eec
WARNING:toil.jobWrapper:Due to failure we are reducing the remaining retry count of job 3b2e3ae6-7a46-4686-99fa-5b0b7dc02eec to 1
```

I haven't deleted the jobstore for this run; it is `aws:us-west-2:fnothaft-toil-jobstore`. I can also send you a capture of the stderr/stdout from the toil leader, if that'd be of interest.

As an aside, there was other anomalous behavior with this job. Although the job was set to request all of the cores and memory from a single machine, it kept on getting scheduled on the same machine as another large job. Additionally, it wasn't clear why this job was failing. It would hang on startup, while a duplicate of the job got scheduled and ran on another node of the cluster. I'm rerunning right now with a different job store and will let you know if the error reproduces.

**Comment by jvivian** [comment]
To add, I ran into a lot of anomalous behavior trying to run something earlier today: resource requirements weren't being respected and jobs would hang anytime I used a returnValue().

**Comment by fnothaft** [comment]
Yeah, I just had another run go wonky. Multiple jobs with mutually exclusive resource requirements got scheduled on a single machine, while another machine went unused while other jobs were waiting to run.

However, I hadn't killed some Docker containers from a prior run (which we really need a strategy for handling), so I killed that job, killed the containers, and am restarting.

**Comment by fnothaft** [comment]
Now, I've got one node where 6 jobs are running (one of which should have requested the full machine), one node where no jobs are running, and two nodes which each have a single scheduled job. Additionally, part of my DAG seems to be blocked...?

**Comment by fnothaft** [comment]
w00t!

**Comment by jvivian** [comment]
@arkal is also having issues, @fnothaft you figure out wtf is going on?

**Comment by fnothaft** [comment]
Nope. No idea. Actually, it looks like the one node is hung.

Also, this is a cluster that I've been running a-OK for a week with no problems until yesterday. Not a new cluster...?

**Comment by jvivian** [comment]
I even tried older versions of Toil (3.1.1) and had the same issue. Other possibilities are CGCloud or AWS itself....but I don't have any evidence whatsoever. Toil versions its dependencies, so I can't think of what other general software dependency/software issue could be causing this.

**Comment by fnothaft** [comment]
I'm mostly perplexed because this cluster has been running OK for a week, and then has been hanging nonstop since yesterday. I'm wondering if there's some sort of silent issue in mesos or on one of the executors?

**Comment by fnothaft** [comment]
The cluster is running successfully now after retrying with a 1 node Spark cluster (this is the ADAM/Toil script --> 1 worker for _totes high performance_), but now that I've gotten to the HaplotypeCaller phase, I should have two jobs running in parallel (ADAM/HapCaller + GATK Best Practices/HapCaller) and I just have one running. On four Toil workers. Wat?

**Comment by hannes-ucsc** [comment]
The retry count not being decremented needs to be looked at. We don't want *any* infinite retries. The reason the worker decrements the retry count and not the leader is to off-load the leader. Unfortunately, this introduces the potential of the worker failing before it has a chance to decrement the count.

The correct pattern is to have the leader decrement the count if the worker failed. This could lead to the count being decremented twice, if the worker fails after decrementing the count but that's still better than an infinite loop.

**Comment by hannes-ucsc** [comment]
We'll address this in the master branch.

**Comment by hannes-ucsc** [comment]
@jvivian says: Haven't seen this in a long time. Feel free to reopen.
